### PR TITLE
Feature#146 add support for automatically updating workflow templates to prepare release

### PR DESCRIPTION
**Pull request by Jannis-Mittenzwei**

### CODE REVIEW

File: test/unit/replace_version_test.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import pytest
+from exasol.toolbox.tools import replace_version
+
+
+def test_retrieve_issue_templates():
```
Comment (Nicoretti) [inline test/unit/replace_version_test.py:5]: try to make a parametrized test out of this.
https://docs.pytest.org/en/7.1.x/how-to/parametrize.html

File: exasol/toolbox/tools/replace_version.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+        _replace(lines, "exasol/python-toolbox/.github/", version)
+        file.seek(0)
+        file.writelines(lines)
+
+
+def _replace(lines: List[str], replace_filter: str, version: str) -> None:
+    for count, line in enumerate(lines):
+        if line.find(replace_filter) != -1:
+            if line.find("@") != -1:
+                lines[count] = line[0:line.index("@")+1] + version + "\n"
```
Comment (Nicoretti) [inline exasol/toolbox/tools/replace_version.py:16]: if you have a loop(s) which select a subset of the iterated items, you usually can simplify the code by creating an iterable over the items you want to take action on. e.g.:

```python
    filtered_lines = ((index, line) for index, line in enumerate(lines) if line.find(replace_filter) != -1)
    filtered_liens = [(index, line) for index, line in filtered_lines if line.find("@") != =1]
    for index, line in filtered_lines:
        lines[index] = line[0:line.index("@")+1] + version + "\n"
```

File: exasol/toolbox/tools/replace_version.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+from pathlib import Path
+from typing import List
+
+def replace_version(template: Path, version: str) -> None:
+    with open(template, "r+", encoding="utf8") as file:
+        lines = file.readlines()
+        _replace(lines, "exasol/python-toolbox/.github/", version)
+        file.seek(0)
+        file.writelines(lines)
```
Comment (Nicoretti) [inline exasol/toolbox/tools/replace_version.py:9]: I'll suggest you do IPO (Input Processing Output) / EVA (Eingabe Verarbeitung Ausgabe).

1. Read the file/lines (open read)
2. Transform data
3.  Write new data (open write)

File: exasol/toolbox/tools/replace_version.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+
+def _replace(lines: List[str], replace_filter: str, version: str) -> None:
```
Comment (Nicoretti) [inline exasol/toolbox/tools/replace_version.py:12]: Usually you want to have the functionality handle a single item, multiple then be handled by adding a loop and potentially wrapping.

e.g.:

```python
def _replace(line, filter, version):
     ....

def _replace_line(lines, filter, version):
      for l in lines:
          _replace(line, filter, version)
```

Additionally you could use "one" function and distinguish using [single_dispatch](https://docs.python.org/3/library/functools.html#functools.singledispatch)
`_replace(...)`, `replace(...)`

File: test/unit/replace_version_test.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+    lines = ["first/line\n", "second/line@0.0.0\n",""]
+    expected = ["first/line\n", "second/line@9.9.9\n",""]
```
Comment (Nicoretti) [inline test/unit/replace_version_test.py:7]: also consider a "negative" test, where input == output because nothing needs to be changed.

**Review by Nicoretti — COMMENTED** [review]
I recommend to create and put this code into `exasol/toolbox/github.py` because this is about updating and dealing with GitHub related things.

File: exasol/toolbox/tools/replace_version.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
+from pathlib import Path
+from typing import List, Tuple
+
+
+def replace_version(template: Path, version: str) -> None:
```
Comment (Nicoretti) [inline exasol/toolbox/tools/replace_version.py:5]:
```suggestion
def update_workflow(template: Path, version: str) -> None:
    """Updates versions of XYZ in GitHub workflow ... """
```

File: exasol/toolbox/tools/replace_version.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    with open(template, encoding="utf-8") as file:
+        input_lines = file.readlines()
```
Comment (Nicoretti) [inline exasol/toolbox/tools/replace_version.py:7]:
```suggestion
    with open(template, encoding="utf-8") as file:
        content = file.readlines()
```

File: exasol/toolbox/tools/replace_version.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+    output_lines = _replace_version(
+        input_lines, "exasol/python-toolbox/.github/", version
+    )
```
Comment (Nicoretti) [inline exasol/toolbox/tools/replace_version.py:11]:
```suggestion
    content = update_versions(
        content, "exasol/python-toolbox/.github/", version
    )
```

File: exasol/toolbox/tools/replace_version.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+    with open(template, "w", encoding="utf-8") as file:
+        file.writelines(output_lines)
```
Comment (Nicoretti) [inline exasol/toolbox/tools/replace_version.py:14]:
```suggestion
    with open(template, "w", encoding="utf-8") as file:
        file.writelines(content)
```

File: exasol/toolbox/tools/replace_version.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        for index, line in filtered_lines
+    ]
+
+
+def _replace_lines(lines: List[str], replace_lines: List[Tuple[int, str]]):
+    output = lines
+    for index, line in replace_lines:
+        output[index] = line
+    return output
+
```
Comment (Nicoretti) [inline exasol/toolbox/tools/replace_version.py:51]:
```suggestion
def is_update_required(line, matcher):
    return line.find(matcher) != -1 and line.find("@") != -1 

def update_version(line, version):
    keep = line[:line.index("@") + 1]
    updated = f"{version}\n"
    return  f"{keep}{updated}"

def update_versions(lines, matcher, version) -> List[str]:
    result = []

    for line in lines:
        if is_update_required(line, matcher):
            line = update_version(line, version)
        result.append(line)

    return result
```

File: test/unit/replace_version_test.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+        (
+            ["path/file@1.2.3\n"],
+            "path/file",
+            "8.9.7",
+            ["path/file@8.9.7\n"]
+        )
+    ]
+)
+def test_retrieve_issue_templates(line, replace_filter, version, expected):
+    actual = replace_version._replace_version(line, replace_filter, version)
```
Comment (Nicoretti) [inline test/unit/replace_version_test.py:28]:
```suggestion
    actual = replace_version.update_version(line, version)
```

File: test/unit/replace_version_test.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
+import pytest
```
Comment (Nicoretti) [inline test/unit/replace_version_test.py:1]: you may also want to add a test for `update_versions` (`update_version` vs `update_versions`).

**Comment by Nicoretti** [comment]
Hi Jannis,

I tweaked the approach a bit. Though it might seem like it contradicts my previous advice on using a filter and a restricted loop, it actually doesn't. The key is we need to process every line, not just select ones. Each line needs a slightly different touch—some we adjust and add, others we just add directly to the output. Dive in, experiment with both your method and this one, and we can hash out the details when I'm back.

Best,
Nico

File: test/unit/replace_version_test.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+
+@pytest.mark.parametrize(
+    "line,version,expected",
+    [
+        ("hallo/world@1.2.3\n", "2.3.4", "hallo/world@2.3.4\n"),
+        ("hallo/world@0.0.0\n", "9.9.9", "hallo/world@9.9.9\n"),
+    ],
+)
+def test_update_required(line, version, expected):
```
Comment (Nicoretti) [inline test/unit/replace_version_test.py:31]:
```suggestion
def test_update_version(line, version, expected):
```

File: noxfile.py [hunk 14]
```diff
@@ -6,3 +6,60 @@
 
 # default actions to be run if nothing is explicitly specified with the -s option
 nox.options.sessions = ["fix"]
+
+
+def main() -> None:
```
Comment (tkilias) [inline noxfile.py:11]: I guess we only need this in this repo

File: noxconfig.py [hunk 15]
```diff
@@ -20,6 +42,7 @@ class Config:
     doc: Path = Path(__file__).parent / "doc"
     version_file: Path = Path(__file__).parent / "exasol" / "toolbox" / "version.py"
     path_filters: Iterable[str] = ("dist", ".eggs", "venv", "metrics-schema")
+    plugins = [UpdateTemplates]
```
Comment (tkilias) [inline noxconfig.py:45]: I guess, we only need this in this repo

**Review by tkilias — APPROVED** [review]

File: noxconfig.py (hunk 15 above)
Comment (Nicoretti) [reply noxconfig.py:45]: Yes, other repos could define their own if needed.

File: noxfile.py (hunk 14 above)
Comment (Nicoretti) [reply noxfile.py:11]: Yes

**Pull request merged**